src/builders/indexable-builder.php

```diff
@@ -0,0 +1,212 @@
+	 * @param Indexable_Hierarchy_Builder         $hierarchy_builder         The hierarchy builder for creating the indexable hierarchy.
+	 */
+	public function __construct(
+		Indexable_Author_Builder $author_builder,
+		Indexable_Post_Builder $post_builder,
+		Indexable_Term_Builder $term_builder,
+		Indexable_Home_Page_Builder $home_page_builder,
+		Indexable_Post_Type_Archive_Builder $post_type_archive_builder,
+		Indexable_Date_Archive_Builder $date_archive_builder,
+		Indexable_System_Page_Builder $system_page_builder,
+		Indexable_Hierarchy_Builder $hierarchy_builder
+	) {
+		$this->author_builder            = $author_builder;
+		$this->post_builder              = $post_builder;
+		$this->term_builder              = $term_builder;
+		$this->home_page_builder         = $home_page_builder;
+		$this->post_type_archive_builder = $post_type_archive_builder;
+		$this->date_archive_builder      = $date_archive_builder;
+		$this->system_page_builder       = $system_page_builder;
+		$this->hierarchy_builder         = $hierarchy_builder;
+	}
+
+	/**
+	 * @required
+	 *
+	 * Sets the indexable repository. Done to avoid circular dependencies.
+	 *
+	 * @param Indexable_Repository $indexable_repository The indexable repository.
+	 */
+	public function set_indexable_repository( Indexable_Repository $indexable_repository ) {
+		$this->indexable_repository = $indexable_repository;
+	}
+
+	/**
+	 * Creates an indexable by its ID and type.
+	 *
+	 * @param int       $object_id   The indexable object ID.
+	 * @param string    $object_type The indexable object type.
+	 * @param Indexable $indexable   Optional. An existing indexable to overwrite.
+	 *
+	 * @return bool|Indexable Instance of indexable.
+	 *
+	 * @throws \Exception If the object_id could not be found.
+	 */
+	public function build_for_id_and_type( $object_id, $object_type, $indexable = false ) {
+		if ( ! $indexable ) {
+			$indexable = $this->indexable_repository->query()->create();
+		}
+
+		switch ( $object_type ) {
+			case 'post':
+				$indexable = $this->post_builder->build( $object_id, $indexable );
+				break;
+			case 'user':
+				$indexable = $this->author_builder->build( $object_id, $indexable );
+				break;
+			case 'term':
+				$indexable = $this->term_builder->build( $object_id, $indexable );
+				break;
+		}
+
+		$indexable->save();
+
+		if ( in_array( $object_type, [ 'post', 'term' ], true ) ) {
+			$this->hierarchy_builder->build( $indexable );
+		}
+
+		return $indexable;
+	}
+
+	/**
+	 * Creates an indexable for the homepage.
+	 *
+	 * @param Indexable $indexable Optional. An existing indexable to overwrite.
+	 *
+	 * @return Indexable The home page indexable.
+	 */
+	public function build_for_home_page( $indexable = false ) {
+		if ( ! $indexable ) {
+			$indexable = $this->indexable_repository->query()->create();
+		}
+		$indexable = $this->home_page_builder->build( $indexable );
```

Comment:
You seems to a save after a build every single time. Have you considered making a method for it: `build_and_save()`. Or something like that?

And maybe you can return that method the indexable directly so you can return that value.